common/lib/xmodule/xmodule/modulestore/tests/test_publish.py

```diff
@@ -139,3 +159,208 @@ def test_publish_draft_delete(self):
             self.draft_mongo.get_item(location)
         self.assertNotIn(other_child_loc, item.children)
         self.assertTrue(self.draft_mongo.has_item(other_child_loc), "Oops, lost moved item")
+
+
+class MixedMongoModulestoreTest(object):
+    modulestore_builder = DRAFT_MODULESTORE_SETUP
+
+
+class MixedSplitModulestoreTest(object):
+    modulestore_builder = SPLIT_MODULESTORE_SETUP
+
+
+class DirectMongoModulestoreTest(object):
+    modulestore_builder = MongoModulestoreBuilder()
+
+
+class UniversalTestSetup(object):
+    """
+    This class exists to test XML import and export between different modulestore
+    classes.
+
+    Requires from subclasses:
+        self.modulestore - modulestore to be tested
+        self.course_key - course key of the test course
+    """
+
+    def _create_course(self, store, course_key):
+        """
+        Create the course that'll be published below. The course has a binary structure, meaning:
+        The course has two chapters (chapter_0 & chapter_1), each of which has two sequentials (seqential_0/1 & sequential_2/3),
+        each of which has two verticals (vertical_0/1 - vertical_6/7), each of which has two units (unit_0/1 - unit_14/15).
+        """
+        def _create_chapter(parent, block_id):
+            return store.create_child(
+                self.user_id, parent.location, 'chapter', block_id=block_id
+            )
+
+        def _create_seq(parent, block_id):
+            return store.create_child(
+                self.user_id, parent.location, 'sequential', block_id=block_id
+            )
+
+        def _create_vert(parent, block_id):
+            return store.create_child(
+                self.user_id, parent.location, 'vertical', block_id=block_id
+            )
+
+        def _create_unit(parent, block_id):
+            return store.create_child(
+                self.user_id, parent.location, 'html', block_id=block_id
+            )
+
+        # Create course.
+        self.course = store.create_course(course_key.org, course_key.course, course_key.run, self.user_id)
+
+        # Create chapters.
+        for idx in xrange(0, 2):
+            setattr(self, 'chapter{}'.format(idx), _create_chapter(self.course, 'chapter_{}'.format(idx)))
+
+        # Create sequentials.
+        for idx in xrange(0, 4):
+            chapter_idx = idx / 2
+            setattr(self, 'seq{}'.format(idx), _create_seq(getattr(self, 'chapter{}'.format(chapter_idx)), 'sequential_{}'.format(idx)))
+
+        # Create verticals.
+        for idx in xrange(0, 8):
+            seq_idx = idx / 2
+            setattr(self, 'vertical{}'.format(idx), _create_vert(getattr(self, 'seq{}'.format(seq_idx)), 'vertical_{}'.format(idx)))
+
+        # Create units.
+        for idx in xrange(0, 16):
+            vert_idx = idx / 2
+            setattr(self, 'unit{}'.format(idx), _create_unit(getattr(self, 'vertical{}'.format(vert_idx)), 'unit_{}'.format(idx)))
+
```

Comment:
Interesting. I was thinking about it as a set of recursive functions, with essentially binary numbers in the item titles.